Plugins/Sycl/src/Seeding/CreateSeedsForGroupSycl.cpp

```diff
@@ -0,0 +1,719 @@
+
+    // Count compatible bottom/top space points per middle space point.
+    std::atomic_uint32_t* deviceCountBotDuplets =
+        cl::sycl::malloc_device<std::atomic_uint32_t>(M, *q);
+    std::atomic_uint32_t* deviceCountTopDuplets =
+        cl::sycl::malloc_device<std::atomic_uint32_t>(M, *q);
+
+    uint32_t* deviceNumTopDuplets = cl::sycl::malloc_device<uint32_t>(M, *q);
+
+    // The limit of compatible bottom [top] space points per middle space point
+    // is B [T]. Temporarily we reserve buffers of this size (M*B and M*T). We
+    // store the indices of bottom [top] space points in bottomSPs [topSPs]. We
+    // move the indices to optimal size vectors for easier indexing.
+    uint32_t* deviceTmpIndBot = cl::sycl::malloc_device<uint32_t>(M * B, *q);
+    uint32_t* deviceTmpIndTop = cl::sycl::malloc_device<uint32_t>(M * T, *q);
+
+    q->memcpy(deviceBottomSPs, bottomSPs.data(),
+              sizeof(detail::DeviceSpacePoint) * (B));
+    q->memcpy(deviceMiddleSPs, middleSPs.data(),
+              sizeof(detail::DeviceSpacePoint) * (M));
+    q->memcpy(deviceTopSPs, topSPs.data(),
+              sizeof(detail::DeviceSpacePoint) * (T));
+    q->memset(deviceCountBotDuplets, 0, M * sizeof(std::atomic_uint32_t));
+    q->memset(deviceCountTopDuplets, 0, M * sizeof(std::atomic_uint32_t));
+
+    //*********************************************//
+    // ********** DUPLET SEARCH - BEGIN ********** //
+    //*********************************************//
+
+    {
+      q->submit([&](cl::sycl::handler& h) {
+        h.parallel_for<duplet_search_bottom_kernel>(
+            cl::sycl::range<2>{M, B}, [=](cl::sycl::id<2> idx) {
+              const auto mid = idx[0];
+              const auto bot = idx[1];
+              // We check whether this thread actually makes sense (within
+              // bounds).
+              if (mid < M && bot < B) {
+                const auto midSP = deviceMiddleSPs[mid];
+                const auto botSP = deviceBottomSPs[bot];
+
+                const auto deltaR = midSP.r - botSP.r;
+                const auto cotTheta = (midSP.z - botSP.z) / deltaR;
+                const auto zOrigin = midSP.z - midSP.r * cotTheta;
+
+                if (!(deltaR < configData.deltaRMin) &&
+                    !(deltaR > configData.deltaRMax) &&
+                    !(cl::sycl::abs(cotTheta) > configData.cotThetaMax) &&
+                    !(zOrigin < configData.collisionRegionMin) &&
+                    !(zOrigin > configData.collisionRegionMax)) {
+                  // We keep counting duplets with atomic variables
+                  const auto ind = deviceCountBotDuplets[mid].fetch_add(1);
+                  deviceTmpIndBot[mid * B + ind] = bot;
+                }
+              }
+            });
+      });
+
+      q->submit([&](cl::sycl::handler& h) {
+        h.parallel_for<duplet_search_top_kernel>(
+            cl::sycl::range<2>{M, B}, [=](cl::sycl::id<2> idx) {
```

Comment:
Just to mark the bug you had found yourself
```suggestion
            cl::sycl::range<2>{M, T}, [=](cl::sycl::id<2> idx) {
```

Reply:
It wasn't really me but thanks,